### ISSUE TITLE
Error installing rar2fs

### ISSUE THREAD

**Issue by burton666** [hasse69/rar2fs#85]
I found a script that where supposed to install rar2fs and rar but I cant get it to work. I get this error-message:

> configure: error: scandir unable to find a working combination
make: *** No targets specified and no makefile found.  Stop.

Can anyone tell me how to modify it to be able to fix the error-message?
I also tried the instructions from the wiki but could not get that to work either.

```
#!/bin/bash
UNRARVER=5.4.5
RAR2FSVER=1.26.0
WORKDIR=`mktemp -d` && cd $WORKDIR

# Get deps
sudo apt-get -q update
sudo apt-get -qy install make libfuse-dev g++

# Get, make and install unrar
wget http://www.rarlab.com/rar/unrarsrc-$UNRARVER.tar.gz
tar zxvf unrarsrc-$UNRARVER.tar.gz
cd unrar
make && sudo make install
make lib && sudo make install-lib
cd ..

# Get, make and install rar2fs
wget https://github.com/hasse69/rar2fs/releases/download/v$RAR2FSVER/rar2fs-$RAR2FSVER.tar.gz
tar zxvf rar2fs-$RAR2FSVER.tar.gz
cd rar2fs-$RAR2FSVER
./configure --with-unrar=../unrar --with-unrar-lib=/usr/lib/
make && sudo make install

# Cleanup
rm -rf $WORKDIR
```

**Comment by hasse69** [comment]
Sounds like something in your build environment or toolchain setup.
Please try this command *before* doing `configure`.

```
autoreconf -fi
```

**Comment by burton666** [comment]
Ok tried that but then I get:

```
 Cannot find unrar source - add alternative source search path using
            --with-unrar and rerun configure.
burton666@burton666-virtual-machine:~/Downloads/rar2fs$ make && sudo make install
make: *** No targets specified and no makefile found.  Stop.
```

**Comment by hasse69** [comment]
Well, you still need to do
```
./configure --with-unrar=../unrar
```
If that is where you installed the unrar source relative to the rar2fs source tree.

**Comment by hasse69** [comment]
Please show what errors you got after following the new instructions.

**Comment by burton666** [comment]
1. I downloaded the latest rar2fs and run "tar -zxvf rar2fs-1.26.0.tar.gz"
2.  cd rar2fs-1.26.0/
3. wget http://www.rarlab.com/rar/unrarsrc-5.5.8.tar.gz
4. tar -zxvf rar2fs-1.26.0.tar.gz
5. tar -zxvf unrarsrc-5.5.8.tar.gz
6. cd unrar
7. make lib
8. sudo make install-lib
9. cd ..
10. autoreconf -f -i
11. ./configure --with-unrar=../unrar
After that is run the last line is: 
"configure: error:
            Cannot find unrar source - add alternative source search path using
            --with-unrar and rerun configure.
"

does not ../unrar mean that it should use the directory /unrar relative to the active directory?

So I tried with: ./configure --with-unrar=/home/burton666/Downloads/rar2fs-1.26.0/unrar but then the last three lines is: 
"checking scandir ARG3 => const struct dirent * ... not ok
checking scandir ARG3 => struct dirent * ... not ok
configure: error: scandir unable to find a working combination
"

**Comment by hasse69** [comment]
> does not ../unrar mean that it should use the directory /unrar relative to the active directory?

Correct. It does. My assumption was that you followed the other instructions/script that for some reason places the unrar source folder on the same level as the rar2fs source folder. With the command sequence you just posted you should not need to use `--with-unrar` at all. The configure script will by default search for the unrar source tree in the same folder as you extracted rar2fs. 

The problem here is the `scandir()` function prototype for some reason cannot be found!? What kind of system is this really? I have never encountered this problem before (or no one has ever told me about it).
Please, can you attach the `config.log` file and also, if possible, attach the `dirent.h` file your system is using. If cross-compiling it should be somewhere in your toolchain root, if this is a native compilation, it is most likely to be found in `/usr/include`.

**Comment by hasse69** [comment]
I think we might have a cause of this failure. If my guess is correct your system is a stricter 64-bit environment that for some reason require different settings than what is expected.
It could be that your system specifically requires the 64-bit variant of the `dirent` structure, something that the detection function in the configure script is not supporting.

**Comment by burton666** [comment]
Ok, the os is Lubuntu and is running on as a virtual machine if that makes any difference.
Is it possible to fix this in some way?
[config.log](https://github.com/hasse69/rar2fs/files/1593281/config.log)
[https://ufile.io/j2sqw](dirent.h)

**Comment by hasse69** [comment]
Yes, it can be fixed. There are many systems out there, impossible to cover them all by myself. This is why I rely on feedback and reports if something is not working. Good catch.
Try the attached patch. You apply it by copying the file to the root of the rar2fs source folder and do:
```
patch -p1 < scandir.patch.txt
```
You should see output like:
```
patching file configure.ac
patching file m4/ax_prototype_scandir.m4
patching file rar2fs.c
```
After patch has been applied, re-run the configuration step:
```
autoreconf -fi
./configure
```
This time is should hopefully pass without errors and you can try:
```
make
make install
```
But I am not sure this is the only problem we will see with this system. Lets hope for the best,

[scandir.patch.txt](https://github.com/hasse69/rar2fs/files/1593364/scandir.patch.txt)

**Comment by burton666** [comment]
```
burton666@burton666-virtual-machine:~/Downloads/rar2fs-1.26.0$ autoreconf -fi
burton666@burton666-virtual-machine:~/Downloads/rar2fs-1.26.0$ ./configure
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... x86_64-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for style of include used by make... GNU
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for gcc option to accept ISO C99... none needed
checking for gcc option to accept ISO Standard C... (cached) none needed
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking dependency style of g++... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking whether byte ordering is bigendian... no
checking for size_t... yes
checking for working alloca.h... yes
checking for alloca... yes
checking errno.h usability... yes
checking errno.h presence... yes
checking for errno.h... yes
checking arpa/inet.h usability... yes
checking arpa/inet.h presence... yes
checking for arpa/inet.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking limits.h usability... yes
checking limits.h presence... yes
checking for limits.h... yes
checking locale.h usability... yes
checking locale.h presence... yes
checking for locale.h... yes
checking malloc.h usability... yes
checking malloc.h presence... yes
checking for malloc.h... yes
checking stddef.h usability... yes
checking stddef.h presence... yes
checking for stddef.h... yes
checking sys/file.h usability... yes
checking sys/file.h presence... yes
checking for sys/file.h... yes
checking sys/mount.h usability... yes
checking sys/mount.h presence... yes
checking for sys/mount.h... yes
checking sys/param.h usability... yes
checking sys/param.h presence... yes
checking for sys/param.h... yes
checking sys/statvfs.h usability... yes
checking sys/statvfs.h presence... yes
checking for sys/statvfs.h... yes
checking syslog.h usability... yes
checking syslog.h presence... yes
checking for syslog.h... yes
checking wchar.h usability... yes
checking wchar.h presence... yes
checking for wchar.h... yes
checking execinfo.h usability... yes
checking execinfo.h presence... yes
checking for execinfo.h... yes
checking ucontext.h usability... yes
checking ucontext.h presence... yes
checking for ucontext.h... yes
checking sched.h usability... yes
checking sched.h presence... yes
checking for sched.h... yes
checking sys/sysmacros.h usability... yes
checking sys/sysmacros.h presence... yes
checking for sys/sysmacros.h... yes
checking sys/xattr.h usability... yes
checking sys/xattr.h presence... yes
checking for sys/xattr.h... yes
checking for dirent.h that defines DIR... yes
checking for library containing opendir... none required
checking for stdbool.h that conforms to C99... no
checking for _Bool... no
checking whether to enable assertions... yes
checking for uid_t in sys/types.h... yes
checking for inline... inline
checking for mode_t... yes
checking for off_t... yes
checking for pid_t... yes
checking for size_t... (cached) yes
checking for ssize_t... yes
checking for struct stat.st_blksize... yes
checking for struct stat.st_gen... no
checking for struct stat.st_atim... yes
checking for struct stat.st_ctim... yes
checking for struct stat.st_mtim... yes
checking for struct stat.st_blocks... yes
checking for struct stat.st_rdev... yes
checking for uint64_t... yes
checking for a working PRIu64... yes
checking return type of signal handlers... void
checking for struct sigaction.sa_sigaction... yes
checking for unistd.h... (cached) yes
checking for working chown... yes
checking vfork.h usability... no
checking vfork.h presence... no
checking for vfork.h... no
checking for fork... yes
checking for vfork... yes
checking for working fork... yes
checking for working vfork... (cached) yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for _LARGEFILE_SOURCE value needed for large files... no
checking whether lstat correctly handles trailing slash... yes
checking for mktime... yes
checking for atexit... yes
checking for dup3... yes
checking for fs_stat_dev... no
checking for ftruncate... yes
checking for getcwd... yes
checking for getpass... yes
checking for lchown... yes
checking for memchr... yes
checking for memmove... yes
checking for memset... yes
checking for mkdir... yes
checking for realpath... yes
checking for rmdir... yes
checking for select... yes
checking for setlocale... yes
checking for strchr... yes
checking for strdup... yes
checking for strerror... yes
checking for strpbrk... yes
checking for strrchr... yes
checking for strstr... yes
checking for strtol... yes
checking for strtoul... yes
checking for utimensat... yes
checking for fdatasync... yes
checking for wcstombs... yes
checking for umask... yes
checking for library containing getxattr... none required
checking for getxattr... yes
checking for lgetxattr... yes
checking for fgetxattr... yes
checking for listxattr... yes
checking for llistxattr... yes
checking for flistxattr... yes
checking for removexattr... yes
checking for lremovexattr... yes
checking for fremovexattr... yes
checking for setxattr... yes
checking for lsetxattr... yes
checking for fsetxattr... yes
checking whether xattr interface takes additional options... no
checking for fmemopen... yes
checking for lockf... yes
checking for flock... yes
checking for stdlib.h... (cached) yes
checking for unistd.h... (cached) yes
checking for sys/param.h... (cached) yes
checking for getpagesize... yes
checking for working mmap... yes
checking for munmap... yes
checking for library containing backtrace... none required
checking for the pthreads library -lpthreads... no
checking whether pthreads work without any flags... no
checking whether pthreads work with -Kthread... no
checking whether pthreads work with -kthread... no
checking for the pthreads library -llthread... no
checking whether pthreads work with -pthread... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... no
checking for PTHREAD_PRIO_INHERIT... yes
checking for main in -lrt... yes
checking for main in -ldl... yes
checking fuse.h usability... yes
checking fuse.h presence... yes
checking for fuse.h... yes
checking for fuse_new in -lfuse4x... no
checking for fuse_new in -lfuse_ino64... no
checking for fuse_new in -lfuse... yes
checking for fuse_version... yes
checking for supported FUSE API version... yes
checking if we need to set -DCYGFUSE... no
checking how to run the C++ preprocessor... g++ -E
checking rar.hpp usability... yes
checking rar.hpp presence... yes
checking for rar.hpp... yes
checking for static linking of UnRAR library... fallback
checking for ./unrar/libunrar.a... yes
checking if unrar library needs -DRAR_SMP... yes
checking for iconv... yes
checking for working iconv... yes
checking for iconv declaration...
         extern size_t iconv (iconv_t cd, char * *inbuf, size_t *inbytesleft, char * *outbuf, size_t *outbytesleft);
checking for commit hash in git repo... no
checking for sched_setaffinity... yes
checking whether cpu_set_t available... yes
checking whether the CPU_SET and CPU_ZERO macros are defined... yes
checking scandir ARG2 => struct dirent ARG3 => const struct dirent ... not ok
checking scandir ARG2 => struct dirent64 ARG3 => const struct dirent ... not ok
checking scandir ARG2 => struct dirent ARG3 => struct dirent ... not ok
checking scandir ARG2 => struct dirent64 ARG3 => struct dirent ... not ok
checking scandir ARG2 => struct dirent ARG3 => const struct dirent64 ... not ok
checking scandir ARG2 => struct dirent64 ARG3 => const struct dirent64 ... not ok
checking scandir ARG2 => struct dirent ARG3 => struct dirent64 ... not ok
checking scandir ARG2 => struct dirent64 ARG3 => struct dirent64 ... not ok
configure: error: scandir unable to find a working combination

```

**Comment by burton666** [comment]
And I did run "patch -p1 < scandir.patch.txt" and got the three lines that I where supposed to.

**Comment by hasse69** [comment]
Well my guess was wrong then :(
Need to look a bit closer at your header file to see what is so special with this system.

**Comment by hasse69** [comment]
The header file looks very similar or even the same as the one on my current system. I have no clue to why it does not work for you :( My best bet is that the file you attached is not actually what is picked up at compile time.

**Comment by burton666** [comment]
Ok, sorry is that the config.log or the dirent.h ?

**Comment by hasse69** [comment]
This is the problem (from your config.log), need to figure out the best way to solve it.
```
conftest.c: In function 'main':
conftest.c:132:3: error: argument 1 null where non-null expected [-Werror=nonnull]
   scandir(dir, namelist, select, alphasort);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from conftest.c:124:0:
/usr/include/dirent.h:255:12: note: in a call to function 'scandir' declared here
 extern int scandir (const char *__restrict __dir,
            ^~~~~~~
conftest.c:132:3: error: argument 2 null where non-null expected [-Werror=nonnull]
   scandir(dir, namelist, select, alphasort);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from conftest.c:124:0:
/usr/include/dirent.h:255:12: note: in a call to function 'scandir' declared here
 extern int scandir (const char *__restrict __dir,
            ^~~~~~~
cc1: all warnings being treated as errors
```

**Comment by hasse69** [comment]
Try this patch instead (you need to revert the previous changes first).
[scandir.patch.txt](https://github.com/hasse69/rar2fs/files/1593522/scandir.patch.txt)

**Comment by burton666** [comment]
Thanks it seams to work now. Last question (Hopefully).
I have mounted a network folder with:

`//diskstation/Main_Storage /media/Main cifs username=Mediaserver,password=PASSWORD,uid=nobody,iocharset=utf8,noperm 0 0`

And can access the share normally.
I then tried this:

`sudo rar2fs /media/Main/Barnfilmer/Movies /home/burton666/media/Barnfilmer`

The first path have nested folders with rar-ed movies and the second path is a empty folder. Have I missed something? The target folder gets converted to a "corrupt file"

**Comment by hasse69** [comment]
I am a bit busy right now, so the answer will be rather short. Will get back to you later with a more detailed answer if needed.

I do not really get why you use sudo? The whole point of fuse is that any user can mount. If you use sudo only root will be able to read the mount point. 
I also see three path given to rar2fs? You should provide the source and the target only. Since you use  CIFS/Samba also provide the allow_other mount option, eg. rar2fs source_dir target_dir -oallow_other.

**Comment by hasse69** [comment]
Sorry, my mail client fooled me a bit, you had two paths there not three, my bad.
According to what you describe I would recommend the following invocation of rar2fs:

```
rar2fs /media/Main/Barnfilmer/Movies /home/burton666/media/Barnfilmer --seek-length=1 -oallow_other
```

**Comment by burton666** [comment]
Thanks for all the help, everything is working now.

**Comment by hasse69** [comment]
Latest version on master should now resolve this issue.

**Comment by swordh** [comment]
Thx @hasse69 , that last scandir.patch.txt solved my issue on ubuntu 17.10!
